Refactor BackForward functionality out of WebPageProxy

sheeparegreat opened the pull request:
#### c4b61a9e31301fe52a054833e08ccd414ebc19cd
<pre>
Refactor BackForward functionality out of WebPageProxy
<a href="https://bugs.webkit.org/show_bug.cgi?id=296513">https://bugs.webkit.org/show_bug.cgi?id=296513</a>
<a href="https://rdar.apple.com/155376413">rdar://155376413</a>

Reviewed by Alex Christensen.

* Source/WebKit/CMakeLists.txt:
* Source/WebKit/DerivedSources-input.xcfilelist:
* Source/WebKit/DerivedSources-output.xcfilelist:
* Source/WebKit/DerivedSources.make:
* Source/WebKit/SourcesCocoa.txt:
* Source/WebKit/UIProcess/ProvisionalPageProxy.cpp:
(WebKit::ProvisionalPageProxy::ProvisionalPageProxy):
(WebKit::ProvisionalPageProxy::initializeWebPage):
(WebKit::ProvisionalPageProxy::didReceiveMessage):
(WebKit::ProvisionalPageProxy::didReceiveSyncMessage):
* Source/WebKit/UIProcess/RemotePageProxy.cpp:
(WebKit::RemotePageProxy::RemotePageProxy):
(WebKit::RemotePageProxy::didReceiveMessage):
(WebKit::RemotePageProxy::didReceiveSyncMessage):
* Source/WebKit/UIProcess/SuspendedPageProxy.cpp:
(WebKit::messageNamesToIgnoreWhileSuspended):
(WebKit::SuspendedPageProxy::SuspendedPageProxy):
(WebKit::SuspendedPageProxy::didReceiveMessage):
* Source/WebKit/UIProcess/SuspendedPageProxy.h:
* Source/WebKit/UIProcess/WebBackForwardList.cpp:
(WebKit::WebBackForwardList::backForwardAddItem):
(WebKit::WebBackForwardList::backForwardAddItemShared):
(WebKit::WebBackForwardList::backForwardSetChildItem):
(WebKit::WebBackForwardList::backForwardClearChildren):
(WebKit::WebBackForwardList::backForwardUpdateItem):
(WebKit::WebBackForwardList::backForwardGoToItem):
(WebKit::WebBackForwardList::backForwardListContainsItem):
(WebKit::WebBackForwardList::backForwardGoToItemShared):
(WebKit::WebBackForwardList::backForwardAllItems):
(WebKit::WebBackForwardList::backForwardItemAtIndex):
(WebKit::WebBackForwardList::backForwardListCounts):
* Source/WebKit/UIProcess/WebBackForwardList.h:
* Source/WebKit/UIProcess/WebBackForwardList.messages.in: Added.
* Source/WebKit/UIProcess/WebPageProxy.cpp:
(WebKit::WebPageProxy::addAllMessageReceivers):
(WebKit::WebPageProxy::backForwardAddItemShared):
(WebKit::WebPageProxy::backForwardGoToItemShared):
(WebKit::WebPageProxy::isRemoteFrameNavigation):
(WebKit::WebPageProxy::backForwardAddItem): Deleted.
(WebKit::WebPageProxy::backForwardSetChildItem): Deleted.
(WebKit::WebPageProxy::backForwardClearChildren): Deleted.
(WebKit::WebPageProxy::backForwardUpdateItem): Deleted.
(WebKit::WebPageProxy::backForwardGoToItem): Deleted.
(WebKit::WebPageProxy::backForwardListContainsItem): Deleted.
(WebKit::WebPageProxy::backForwardAllItems): Deleted.
(WebKit::WebPageProxy::backForwardItemAtIndex): Deleted.
(WebKit::WebPageProxy::backForwardListCounts): Deleted.
* Source/WebKit/UIProcess/WebPageProxy.h:
* Source/WebKit/UIProcess/WebPageProxy.messages.in:
* Source/WebKit/UIProcess/WebPageProxyMessageReceiverRegistration.cpp:
(WebKit::WebPageProxyMessageReceiverRegistration::startReceivingMessages):
(WebKit::WebPageProxyMessageReceiverRegistration::stopReceivingMessages):
(WebKit::WebPageProxyMessageReceiverRegistration::transferMessageReceivingFrom):
* Source/WebKit/UIProcess/WebPageProxyMessageReceiverRegistration.h:
* Source/WebKit/WebKit.xcodeproj/project.pbxproj:
* Source/WebKit/WebProcess/WebPage/WebBackForwardListProxy.cpp:
(WebKit::WebBackForwardListProxy::addItem):
(WebKit::WebBackForwardListProxy::setChildItem):
(WebKit::WebBackForwardListProxy::goToItem):
(WebKit::WebBackForwardListProxy::allItems):
(WebKit::WebBackForwardListProxy::itemAtIndex):
(WebKit::WebBackForwardListProxy::containsItem const):
(WebKit::WebBackForwardListProxy::cacheListCountsIfNecessary const):
* Source/WebKit/WebProcess/WebPage/WebHistoryItemClient.cpp:
(WebKit::WebHistoryItemClient::historyItemChanged):
(WebKit::WebHistoryItemClient::clearChildren const):

Canonical link: <a href="https://commits.webkit.org/300863@main">https://commits.webkit.org/300863@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/97f5ab76c53034f1136d00e72d4552acd4b4ac82

| Misc | iOS, visionOS, tvOS & watchOS  | macOS  | Linux |  Windows | Apple Internal |
| ----- | ---------------------- | ------- |  ----- |  --------- | ------ |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/123654 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/131/builds/43369 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/138/builds/34065 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/130428 "Built successfully") | [✅ 🛠 win](https://ews-build.webkit.org/#/builders/59/builds/75799 "Built successfully") | [✅ 🛠 ios-apple](https://ews-bridge.webkit.apple.com/builds/sw/T-276/92699e7b-e2c4-41dd-8b79-0fc7c00d5b52/93dc5988-b3e0-4222-a99f-c72e1ba14e3f) 
| | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/130/builds/44092 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/123/builds/51963 "Built successfully") | [✅ 🧪 wpe-wk2](https://ews-build.webkit.org/#/builders/34/builds/94025 "Passed tests") | [✅ 🧪 win-tests](https://ews-build.webkit.org/#/builders/60/builds/62410 "Passed tests") | [✅ 🛠 mac-apple](https://ews-bridge.webkit.apple.com/builds/sw/T-276/92699e7b-e2c4-41dd-8b79-0fc7c00d5b52/2bea08f8-152b-4898-b3b6-773162d7a73c) 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/126607 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/132/builds/35121 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/110621 "Passed tests") | [✅ 🧪 api-wpe](https://ews-build.webkit.org/#/builders/41/builds/74626 "Passed tests") | | [✅ 🛠 vision-apple](https://ews-bridge.webkit.apple.com/builds/sw/T-276/92699e7b-e2c4-41dd-8b79-0fc7c00d5b52/e7787ac4-d441-44c8-a33a-64477041341c) 
| | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/133/builds/34084 "Passed tests") | | [✅ 🛠 wpe-cairo](https://ews-build.webkit.org/#/builders/65/builds/73901 "Built successfully") | | 
| | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/104842 "Passed tests") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/136/builds/29003 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/133117 "Built successfully") | | 
| | [✅ 🛠 vision](https://ews-build.webkit.org/#/builders/128/builds/50605 "Built successfully") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/122/builds/38527 "Passed tests") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/1/builds/102496 "Passed tests") | | 
| | [✅ 🛠 vision-sim](https://ews-build.webkit.org/#/builders/121/builds/50980 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/8/builds/106842 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/102337 "Passed tests") | | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/19/builds/26123 "Built successfully and passed tests") | [✅ 🧪 vision-wk2](https://ews-build.webkit.org/#/builders/126/builds/47688 "Passed tests") | [✅ 🧪 mac-intel-wk2](https://ews-build.webkit.org/#/builders/137/builds/25930 "Passed tests") | [✅ 🛠 playstation](https://ews-build.webkit.org/#/builders/134/builds/47421 "Built successfully") | | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/127/builds/50459 "Built successfully") | [✅ 🛠 mac-safer-cpp](https://ews-build.webkit.org/#/builders/120/builds/56221 "Built successfully") | | | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/125/builds/49933 "Built successfully") | | | | 
| | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/129/builds/53280 "Built successfully") | | | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/124/builds/51608 "Built successfully") | | | | 
<!--EWS-Status-Bubble-End-->

adetaylor commented on Source/WebKit/UIProcess/ProvisionalPageProxy.cpp:
```diff
@@ -844,8 +854,12 @@ void ProvisionalPageProxy::didReceiveSyncMessage(IPC::Connection& connection, IP
-    if (page)
-        page->didReceiveSyncMessage(connection, decoder, replyEncoder);
+    if (page) {
+        if (decoder.messageReceiverName() == Messages::WebBackForwardList::messageReceiverName())
```
I don't know if you'd consider it _cleaner_ but it would almost certainly be less code to put this check inside `WebPageProxy::didReceiveMessage` (and equivalent for sync messages) and choose to forward messages onto `WebBackForwardList` at that point. This I think would remove the need for some of these changes in `ProvisionalPageProxy.cpp` and `RemotePageProxy.cpp`.

adetaylor commented on Source/WebKit/UIProcess/WebBackForwardListMessageReceiverRegistration.h:
```diff
@@ -0,0 +1,54 @@
+
+class WebProcessProxy;
+
+class WebBackForwardListMessageReceiverRegistration {
```
Did you consider instead adjusting the existing `WebPageProxyMessageReceiverRegistration` to accept references to two `IPC::MessageReceiver`s instead, and removing this new class? I think that would result in less code duplication and be net simpler

achristensen07 replied:
I agree, and this is a pattern we're probably going to do with more and more objects, so it'll get messier and messier with this approach.

sheeparegreat replied:
Done

sheeparegreat committed:
Refactor BackForward functionality out of WebPageProxy
https://bugs.webkit.org/show_bug.cgi?id=296513
rdar://155376413

Reviewed by Alex Christensen.

* Source/WebKit/CMakeLists.txt:
* Source/WebKit/DerivedSources-input.xcfilelist:
* Source/WebKit/DerivedSources-output.xcfilelist:
* Source/WebKit/DerivedSources.make:
* Source/WebKit/SourcesCocoa.txt:
* Source/WebKit/UIProcess/ProvisionalPageProxy.cpp:
(WebKit::ProvisionalPageProxy::ProvisionalPageProxy):
(WebKit::ProvisionalPageProxy::initializeWebPage):
(WebKit::ProvisionalPageProxy::didReceiveMessage):
(WebKit::ProvisionalPageProxy::didReceiveSyncMessage):
* Source/WebKit/UIProcess/RemotePageProxy.cpp:
(WebKit::RemotePageProxy::RemotePageProxy):
(WebKit::RemotePageProxy::didReceiveMessage):
(WebKit::RemotePageProxy::didReceiveSyncMessage):
* Source/WebKit/UIProcess/SuspendedPageProxy.cpp:
(WebKit::messageNamesToIgnoreWhileSuspended):
(WebKit::SuspendedPageProxy::SuspendedPageProxy):
(WebKit::SuspendedPageProxy::didReceiveMessage):
* Source/WebKit/UIProcess/SuspendedPageProxy.h:
* Source/WebKit/UIProcess/WebBackForwardList.cpp:
(WebKit::WebBackForwardList::backForwardAddItem):
(WebKit::WebBackForwardList::backForwardAddItemShared):
(WebKit::WebBackForwardList::backForwardSetChildItem):
(WebKit::WebBackForwardList::backForwardClearChildren):
(WebKit::WebBackForwardList::backForwardUpdateItem):
(WebKit::WebBackForwardList::backForwardGoToItem):
(WebKit::WebBackForwardList::backForwardListContainsItem):
(WebKit::WebBackForwardList::backForwardGoToItemShared):
(WebKit::WebBackForwardList::backForwardAllItems):
(WebKit::WebBackForwardList::backForwardItemAtIndex):
(WebKit::WebBackForwardList::backForwardListCounts):
* Source/WebKit/UIProcess/WebBackForwardList.h:
* Source/WebKit/UIProcess/WebBackForwardList.messages.in: Added.
* Source/WebKit/UIProcess/WebPageProxy.cpp:
(WebKit::WebPageProxy::addAllMessageReceivers):
(WebKit::WebPageProxy::backForwardAddItemShared):
(WebKit::WebPageProxy::backForwardGoToItemShared):
(WebKit::WebPageProxy::isRemoteFrameNavigation):
(WebKit::WebPageProxy::backForwardAddItem): Deleted.
(WebKit::WebPageProxy::backForwardSetChildItem): Deleted.
(WebKit::WebPageProxy::backForwardClearChildren): Deleted.
(WebKit::WebPageProxy::backForwardUpdateItem): Deleted.
(WebKit::WebPageProxy::backForwardGoToItem): Deleted.
(WebKit::WebPageProxy::backForwardListContainsItem): Deleted.
(WebKit::WebPageProxy::backForwardAllItems): Deleted.
(WebKit::WebPageProxy::backForwardItemAtIndex): Deleted.
(WebKit::WebPageProxy::backForwardListCounts): Deleted.
* Source/WebKit/UIProcess/WebPageProxy.h:
* Source/WebKit/UIProcess/WebPageProxy.messages.in:
* Source/WebKit/UIProcess/WebPageProxyMessageReceiverRegistration.cpp:
(WebKit::WebPageProxyMessageReceiverRegistration::startReceivingMessages):
(WebKit::WebPageProxyMessageReceiverRegistration::stopReceivingMessages):
(WebKit::WebPageProxyMessageReceiverRegistration::transferMessageReceivingFrom):
* Source/WebKit/UIProcess/WebPageProxyMessageReceiverRegistration.h:
* Source/WebKit/WebKit.xcodeproj/project.pbxproj:
* Source/WebKit/WebProcess/WebPage/WebBackForwardListProxy.cpp:
(WebKit::WebBackForwardListProxy::addItem):
(WebKit::WebBackForwardListProxy::setChildItem):
(WebKit::WebBackForwardListProxy::goToItem):
(WebKit::WebBackForwardListProxy::allItems):
(WebKit::WebBackForwardListProxy::itemAtIndex):
(WebKit::WebBackForwardListProxy::containsItem const):
(WebKit::WebBackForwardListProxy::cacheListCountsIfNecessary const):
* Source/WebKit/WebProcess/WebPage/WebHistoryItemClient.cpp:
(WebKit::WebHistoryItemClient::historyItemChanged):
(WebKit::WebHistoryItemClient::clearChildren const):

Canonical link: https://commits.webkit.org/300863@main
